### PR TITLE
Add support for SwiftWasm with CI and tests

### PR DESCRIPTION
**Pull request by MaxDesiatov**
WebAssembly support for atomics and multi-threading isn't fully standardized yet, and it not supported in SwiftWasm at the moment. Because of this `Dispatch` is unavailable, and all Combine-related Foundation stuff is unavailable too. Tests related to this are disabled. Locking functions are replaced with no-op shims.

### CODE REVIEW

**Comment by MaxDesiatov** [comment]
Argh, Swift older than 5.3 doesn't support `#if os(WASI)` checks.

@broadwaylamb WDYT about dropping support for pre-5.3 Swift versions? You can still deploy with Swift 5.3 pretty far back, and support old versions of operating systems we already support. And not that 5.3 was a particularly breaking release, I hope it's easy enough if not seamless for people to upgrade.

**Comment by broadwaylamb** [comment]
> Argh, Swift older than 5.3 doesn't support `#if os(WASI)` checks.
> 
> @broadwaylamb WDYT about dropping support for pre-5.3 Swift versions? You can still deploy with Swift 5.3 pretty far back, and support old versions of operating systems we already support. And not that 5.3 was a particularly breaking release, I hope it's easy enough if not seamless for people to upgrade.

You can do this:

```swift
#if swift(>=5.3)
#if os(WASI)

#endif
#endif
```

I don't want to drop support for older Swift versions because if we do, we won't be able to run tests with Xcode 9.3, which is the last Xcode version that can build for 32 bit simulators. And I'd like to be able to test OpenCombine on 32 bit platforms.

**Comment by MaxDesiatov** [comment]
Right, the problem is more about `#if !os(WASI)`. Positive OS conditons can be hidden with `#if swift(>=5.3)`, but I need to exclude some tests, that's why the negative `!os(WASI)` condition is needed too. Is there a way to work around that?

File: Sources/OpenCombine/Helpers/Locking.swift [hunk 1]
```diff
@@ -9,5 +9,22 @@
 import COpenCombineHelpers
 #endif
 
+#if os(WASI)
+struct UnfairLock {
+    public static func allocate() -> UnfairLock { return Self.init() }
+    public func lock() {}
+    public func unlock() {}
+    public func assertOwner() {}
+    public func deallocate() {}
```
Comment (OpenCombineBot) [inline Sources/OpenCombine/Helpers/Locking.swift:18]: - :no_entry_sign: Ensure definitions have a lower access control level than their enclosing parent (`lower_acl_than_parent`)

File: Sources/OpenCombine/Helpers/Locking.swift [hunk 2]
```diff
@@ -9,5 +9,22 @@
 import COpenCombineHelpers
 #endif
 
+#if os(WASI)
+struct UnfairLock {
```
Comment (OpenCombineBot) [inline Sources/OpenCombine/Helpers/Locking.swift:13]: - :no_entry_sign: All declarations should specify Access Control Level keywords explicitly. (`explicit_acl`)
- :no_entry_sign: Top-level declarations should specify Access Control Level keywords explicitly. (`explicit_top_level_acl`)

**Comment by broadwaylamb** [comment]
Oh, that.

At times like this I wish I quit programming.

I'll figure something out.

**Comment by MaxDesiatov** [comment]
BTW, if you care about 32-bit, WebAssembly is currenly 32-bit only 🙂

**Comment by broadwaylamb** [comment]
Here's an idea. 

Instead of writing `#if !os(WASI)` let's use `#if canImport(Darwin) || os(Linux)`, just like you do in `Package.swift`.

File: Sources/OpenCombine/Helpers/Locking.swift [hunk 3]
```diff
@@ -9,5 +9,22 @@
...
+struct UnfairRecursiveLock {
+    public static func allocate() -> UnfairRecursiveLock { return Self.init() }
+    public func lock() {}
+    public func unlock() {}
+    public func deallocate() {}
+}
+#else
 internal typealias UnfairLock = __UnfairLock
 internal typealias UnfairRecursiveLock = __UnfairRecursiveLock
+#endif
```
Comment (broadwaylamb) [inline Sources/OpenCombine/Helpers/Locking.swift]:
```suggestion
#if swift(>=5.3)
#if os(WASI)
internal struct __UnfairLock {
    internal static func allocate() -> UnfairLock { return .init() }
    internal func lock() {}
    internal func unlock() {}
    internal func assertOwner() {}
    internal func deallocate() {}
}

internal struct __UnfairRecursiveLock {
    internal static func allocate() -> UnfairRecursiveLock { return .init() }
    internal func lock() {}
    internal func unlock() {}
    internal func deallocate() {}
}
#endif // os(WASI)
#endif // swift(>=5.3)

internal typealias UnfairLock = __UnfairLock
internal typealias UnfairRecursiveLock = __UnfairRecursiveLock
```

File: Package.swift [hunk 4]
```diff
@@ -11,15 +11,34 @@ let package = Package(
     ],
     targets: [
         .target(name: "COpenCombineHelpers"),
-        .target(name: "OpenCombine", dependencies: ["COpenCombineHelpers"]),
+        .target(
+          name: "OpenCombine", 
+          dependencies: [
+            .target(name: "COpenCombineHelpers", condition: .when(platforms: [.macOS, .linux]))
```
Comment (broadwaylamb) [inline Package.swift]: Let's not forget non-desktop systems here and below.

```suggestion
            .target(name: "COpenCombineHelpers", condition: .when(platforms: [.macOS, .iOS, .watchOS, .tvOS, .linux]))
```

File: Sources/OpenCombine/CombineIdentifier.swift [hunk 5]
```diff
@@ -9,6 +9,14 @@
 import COpenCombineHelpers
 #endif
 
+#if os(WASI)
+private var __identifier: UInt64 = 0
+func __nextCombineIdentifier() -> UInt64 {
+    __identifier += 1
+    return __identifier
```
Comment (broadwaylamb) [inline Sources/OpenCombine/CombineIdentifier.swift]: For consistency with another implementation of this function:
```suggestion
    defer { __identifier += 1 }
    return __identifier
```

**Review by broadwaylamb — COMMENTED** [review]
Great work!

**Comment by MaxDesiatov** [comment]
Turns out that `os(WASI)` is not a compiler error with older versions, just a warning. The error was caused by missing `Package@swift-5.1.swift` and `Package@swift-5.2.swift`.

**Comment by broadwaylamb** [comment]
> Turns out that `os(WASI)` is not a compiler error with older versions, just a warning. The error was caused by missing `Package@swift-5.1.swift` and `Package@swift-5.2.swift`.

Yeah, but I'd like to avoid warnings as well.

File: Package.swift [hunk 6]
```diff
@@ -11,15 +11,34 @@ let package = Package(
...
+          ]
+        ),
+        .testTarget(
+          name: "OpenCombineTests",
+          dependencies: [
+            "OpenCombine",
+            .target(name: "OpenCombineDispatch", condition: .when(platforms: [.macOS, .linux])),
+            .target(name: "OpenCombineFoundation", condition: .when(platforms: [.macOS, .linux])),
+          ],
```
Comment (broadwaylamb) [inline Package.swift]:
```suggestion
        .target(
          name: "OpenCombineFoundation", 
          dependencies: [
            "OpenCombine",
            .target(name: "COpenCombineHelpers", condition: .when(platforms: [.macOS, .iOS, .watchOS, .tvOS, .linux]))
          ]
        ),
        .testTarget(
          name: "OpenCombineTests",
          dependencies: [
            "OpenCombine",
            .target(name: "OpenCombineDispatch", condition: .when(platforms: [.macOS, .iOS, .watchOS, .tvOS, .linux])),
            .target(name: "OpenCombineFoundation", condition: .when(platforms: [.macOS, .iOS, .watchOS, .tvOS, .linux])),
          ],
```

File: Package@swift-5.1.swift [hunk 7]
```diff
@@ -0,0 +1,25 @@
+// swift-tools-version:5.0
```
Comment (broadwaylamb) [inline Package@swift-5.1.swift]:
```suggestion
// swift-tools-version:5.1
```

File: Package@swift-5.2.swift [hunk 8]
```diff
@@ -0,0 +1,25 @@
+// swift-tools-version:5.0
```
Comment (broadwaylamb) [inline Package@swift-5.2.swift]:
```suggestion
// swift-tools-version:5.2
```

File: Tests/OpenCombineTests/Helpers/AssertCrashes.swift [hunk 9]
```diff
@@ -29,7 +29,7 @@ extension XCTest {
     // Taken from swift-corelibs-foundation and slightly modified for OpenCombine
     @available(macOS 10.13, iOS 8.0, *)
     func assertCrashes(within block: () throws -> Void) rethrows {
-#if !Xcode && !os(iOS) && !os(watchOS) && !os(tvOS)
+#if !Xcode && !os(iOS) && !os(watchOS) && !os(tvOS) && !os(WASI)
```
Comment (MaxDesiatov) [inline Tests/OpenCombineTests/Helpers/AssertCrashes.swift]: @broadwaylamb what would you like me to do with this one?

Comment (broadwaylamb) [reply Tests/OpenCombineTests/Helpers/AssertCrashes.swift]:
```swift
#if os(macOS) && !Xcode || os(Linux)
```

**Comment by MaxDesiatov** [comment]
I can't reproduce this error with Swift 5.2 on Linux locally, have you seen it before?

```
[100/114] Compiling OpenCombineFoundation Violations.swiftwift2K[2K<unknown>:0: error: missing submodule 'CFURLSessionInterface'
/root/project/Sources/OpenCombineFoundation/URLSession.swift:11:8: error: missing required module 'CFURLSessionInterface'
import FoundationNetworking
       ^
[101/114] Compiling OpenCombineFoundation JSONEncoder.swift<unknown>:0: error: missing submodule 'CFURLSessionInterface'
/root/project/Sources/OpenCombineFoundation/URLSession.swift:11:8: error: missing required module 'CFURLSessionInterface'
import FoundationNetworking
       ^
[102/114] Compiling OpenCombineFoundation RunLoop+Scheduler.swift<unknown>:0: error: missing submodule 'CFURLSessionInterface'
/root/project/Sources/OpenCombineFoundation/URLSession.swift:11:8: error: missing required module 'CFURLSessionInterface'
import FoundationNetworking
       ^
[106/114] Compiling OpenCombineFoundation OperationQueue+Scheduler.swift<unknown>:0: error: missing submodule 'CFURLSessionInterface'
/root/project/Sources/OpenCombineFoundation/URLSession.swift:11:8: error: missing required module 'CFURLSessionInterface'
import FoundationNetworking
       ^
```

**Comment by broadwaylamb** [comment]
I've rerun the tests, it's fine now.

File: Sources/OpenCombine/Helpers/Locking.swift [hunk 10]
```diff
@@ -9,5 +9,24 @@
 import COpenCombineHelpers
 #endif
 
+#if swift(>=5.3)
+#if os(WASI)
+internal struct __UnfairLock {
```
Comment (OpenCombineBot) [inline Sources/OpenCombine/Helpers/Locking.swift:14]: - :no_entry_sign: Type name should only contain alphanumeric characters: '__UnfairLock' (`type_name`)

Comment (broadwaylamb) [inline Sources/OpenCombine/Helpers/Locking.swift]:
```suggestion
internal struct __UnfairLock { // swiftlint:disable:this type_name
```

File: Sources/OpenCombine/Helpers/Locking.swift [hunk 11]
```diff
@@ -5,9 +5,30 @@
...
+    internal static func allocate() -> UnfairLock { return .init() }
+    internal func lock() {}
+    internal func unlock() {}
+    internal func assertOwner() {}
+    internal func deallocate() {}
+}
+
+internal struct __UnfairRecursiveLock {
```
Comment (broadwaylamb) [inline Sources/OpenCombine/Helpers/Locking.swift]:
```suggestion
internal struct __UnfairRecursiveLock { // swiftlint:disable:this type_name
```

File: Tests/OpenCombineTests/SubscribersTests/SubscribersDemandTests.swift [hunk 12]
```diff
@@ -315,6 +316,8 @@ final class SubscribersDemandTests: XCTestCase {
 
         XCTAssertEqual(decodedIllFormedTooBig.value.description, "unlimited")
     }
+
+    #endif // canImport(Darwin) || os(Linux)
```
Comment (broadwaylamb) [inline Tests/OpenCombineTests/SubscribersTests/SubscribersDemandTests.swift]:
```suggestion
#endif // canImport(Darwin) || os(Linux)
```

File: Sources/OpenCombine/CombineIdentifier.swift [hunk 13]
```diff
@@ -9,6 +9,16 @@
 import COpenCombineHelpers
 #endif
 
+#if swift(>=5.3)
+#if os(WASI)
+private var __identifier: UInt64 = 0
+func __nextCombineIdentifier() -> UInt64 {
```
Comment (OpenCombineBot) [inline Sources/OpenCombine/CombineIdentifier.swift:15]: - :no_entry_sign: All declarations should specify Access Control Level keywords explicitly. (`explicit_acl`)
- :no_entry_sign: Let and var should be separated from other statements by a blank line. (`let_var_whitespace`)
- :no_entry_sign: Top-level declarations should specify Access Control Level keywords explicitly. (`explicit_top_level_acl`)

File: Package.swift [hunk 14]
```diff
@@ -11,15 +11,34 @@ let package = Package(
...
+        .target(
+          name: "OpenCombine", 
+          dependencies: [
+            .target(name: "COpenCombineHelpers", condition: .when(platforms: [.macOS, .iOS, .watchOS, .tvOS, .linux]))
+          ],
+          exclude: [
+            "Publishers/Publishers.Encode.swift.gyb",
+            "Publishers/Publishers.MapKeyPath.swift.gyb",
+            "Publishers/Publishers.Catch.swift.gyb"
+          ]
```
Comment (broadwaylamb) [inline Package.swift]: Could you please explain why is this necessary?

**Review by broadwaylamb — COMMENTED** [review]
Wait. I've just learned that `wasi` is [a supported platform in the PackageDescription API](https://github.com/apple/swift-package-manager/blob/2bec212061295719620c7f4cf2d2e257a95aab39/Sources/PackageDescription/SupportedPlatforms.swift#L47).

Can you add the following parameter to the `OpenCombine` and `OpenCombineTests` targets?

```swift
swiftSettings: [.define("WASI", .when(platforms: [.wasi]))]
```

This way instead of `#if os(WASI)` we can write just `#if WASI` in the source files.

Sorry I didn't realize it before.

File: Package.swift (hunk 14 above)
Comment (MaxDesiatov) [reply Package.swift]: Swift 5.3 package manifests require you either to add all files in the `Sources` as SwiftPM resources, or to exclude them explicitly.

**Comment by MaxDesiatov** [comment]
Writing `#if !WASI` is still not possible, is it? If not, that doesn't really solve the problem with the negative `#if !os(WASI)` platform checks.

**Comment by broadwaylamb** [comment]
Why not? AFAIU, it should be possible. You won't even have to surround it with `#if swift(>=5.3)`.

**Comment by MaxDesiatov** [comment]
You're right, it does work. I always thought that Swift `#if` is fully equivalent to C `#ifdef`, but wasn't sure whether `#if !` is valid and if it translates to `#ifndef`.

**Comment by MaxDesiatov** [comment]
The iOS tests job also fails in `master` with `Failed to download iOS 9.3 Simulator` error.

**Comment by broadwaylamb** [comment]
> The iOS tests job also fails in `master` with `Failed to download iOS 9.3 Simulator` error.

Unfortunately, this happens sometimes.

File: Tests/OpenCombineTests/FoundationTests/OperationQueueSchedulerTests.swift [hunk 15]
```diff
@@ -5,6 +5,10 @@
 //  Created by Sergej Jaskiewicz on 14.06.2020.
 //
 
+// OperationQueue has serious bugs in swift-corelibs-foundation prior to Swift 5.3.
+// (see https://github.com/apple/swift-corelibs-foundation/pull/2779)
+#if canImport(Darwin) || (swift(>=5.3) && os(Linux)) // TEST_DISCOVERY_CONDITION
+
```
Comment (broadwaylamb) [inline Tests/OpenCombineTests/FoundationTests/OperationQueueSchedulerTests.swift]:
```suggestion
// OperationQueue has serious bugs in swift-corelibs-foundation prior to Swift 5.3.
// (see https://github.com/apple/swift-corelibs-foundation/pull/2779)
#if canImport(Darwin) || swift(>=5.3) && !WASI // TEST_DISCOVERY_CONDITION

```

Please also update the comment after `#endif`.

File: Tests/OpenCombineTests/FoundationTests/PropertyListDecoderTests.swift [hunk 16]
```diff
@@ -5,6 +5,10 @@
 //  Created by Sergej Jaskiewicz on 10.12.2019.
 //
 
+// PropertyListEncoder and PropertyListDecoder are unavailable in
+// swift-corelibs-foundation prior to Swift 5.1.
+#if canImport(Darwin) || swift(>=5.1) && os(Linux) // TEST_DISCOVERY_CONDITION
+
```
Comment (broadwaylamb) [inline Tests/OpenCombineTests/FoundationTests/PropertyListDecoderTests.swift]:
```suggestion
// PropertyListEncoder and PropertyListDecoder are unavailable in
// swift-corelibs-foundation prior to Swift 5.1.
#if canImport(Darwin) || swift(>=5.1) && !WASI // TEST_DISCOVERY_CONDITION

```

Please also update the comment after `#endif.`

File: Tests/OpenCombineTests/FoundationTests/PropertyListEncoderTests.swift [hunk 17]
```diff
@@ -5,6 +5,10 @@
 //  Created by Sergej Jaskiewicz on 10.12.2019.
 //
 
+// PropertyListEncoder and PropertyListDecoder are unavailable in
+// swift-corelibs-foundation prior to Swift 5.1.
+#if canImport(Darwin) || (swift(>=5.1) && os(Linux)) // TEST_DISCOVERY_CONDITION
+
```
Comment (broadwaylamb) [inline Tests/OpenCombineTests/FoundationTests/PropertyListEncoderTests.swift]:
```suggestion
// PropertyListEncoder and PropertyListDecoder are unavailable in
// swift-corelibs-foundation prior to Swift 5.1.
#if canImport(Darwin) || swift(>=5.1) && !WASI // TEST_DISCOVERY_CONDITION

```

Please also update the comment after `#endif`.

File: Tests/OpenCombineTests/Helpers/AssertCrashes.swift [hunk 18]
```diff
@@ -29,7 +29,7 @@ extension XCTest {
     // Taken from swift-corelibs-foundation and slightly modified for OpenCombine
     @available(macOS 10.13, iOS 8.0, *)
     func assertCrashes(within block: () throws -> Void) rethrows {
-#if !Xcode && !os(iOS) && !os(watchOS) && !os(tvOS)
+#if os(macOS) && !Xcode || os(Linux)
```
Comment (broadwaylamb) [inline Tests/OpenCombineTests/Helpers/AssertCrashes.swift]:
```suggestion
#if !Xcode && !os(iOS) && !os(watchOS) && !os(tvOS) && !WASI
```

**Review by broadwaylamb — APPROVED** [review]
Awesome, thank you!

One last question: why did you disable the `print()` operator?

**Comment by MaxDesiatov** [comment]
I only disabled `PrintTests.swift` as it relied on `Atomic` and `race`.

**Comment by MaxDesiatov** [comment]
Would you prefer "Squash and merge" or "Rebase and merge" here?

**Comment by broadwaylamb** [comment]
> I only disabled `PrintTests.swift` as it relied on `Atomic` and `race`.

I see. Can I ask you to enable those tests, but tweak the implementation of `Atomic` to not use `NSLock` on Wasm, and the implementation of the `race` function to just use a `for` loop instead of `DispatchQueue.concurrentPerform(iterations:)` on Wasm?

> Would you prefer "Squash and merge" or "Rebase and merge" here?

The former.

File: Tests/OpenCombineTests/Helpers/TestingThreadSafety.swift [hunk 19]
```diff
@@ -6,42 +6,63 @@
...
 import Dispatch
+#endif
+
 import Foundation
 import XCTest
 
 func race(times: Int = 100, _ bodies: () -> Void...) {
+    #if WASI
+    for body in bodies {
+        for _ in 0..<times {
```
Comment (MaxDesiatov) [inline Tests/OpenCombineTests/Helpers/TestingThreadSafety.swift:18]: Not sure if this inner loop is really doing anything useful if there's no actual race occurring?

Comment (broadwaylamb) [reply Tests/OpenCombineTests/Helpers/TestingThreadSafety.swift:18]: Otherwise the semantics would be different.

**Review by broadwaylamb — APPROVED** [review]
